enhancements/project-request-limit.md

```diff
@@ -0,0 +1,293 @@
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+The `requestlimit.project.openshift.io/ProjectRequestLimit` admission plugin is used to impose a limit on the number of
+self-provisioned project(s) requested by a given user. In 3.x, the limit can be specified via the master configuration file.
+It doesn't fit well with our targeted configuration which aims to avoid adding lots of intricately documented knobs. 
+This is why this admission plug-in is disabled in OpenShift 4.x. 
+
+Instead of wiring the admission plugin via the `kube-apiserver` operator, we can create a validating admission webhook 
+based on the [generic-admission-server](https://github.com/openshift/generic-admission-server) and distribute it via
+OperatorHub/OLM.
+
+## Motivation
+* Some customers use the `ProjectRequestLimit` plugin in 3.x and thus are blocked on their path to OpenShift 4.x. Enabling
+ it on 4.x will unblock these customers.
+* The `kube-apiserver` is designed to be extended using validating/mutating admission webhook(s), we have the
+technology to easily build one, and we have the ability to create a simple operator to manage it.
+* The cluster admin can optionally enable the validating webhook on a cluster and configure limits via a custom
+resource. This will sever the "legacy" link between the plugin and the `kube-apiserver`.
+* We can provide the customer with seamless install and automatic upgrades by shipping the operator via OperatorHub 
+  independent of OpenShift release cycle.   
+
+
+### Goals
+1. Enable the `requestlimit.project.openshift.io/ProjectRequestLimit` admission plugin.
+2. Use existing extension points, libraries, and installation mechanisms in the manner we would recommend to external teams.
+3. Have a fairly straightforward way to install and enable this admission plugin.
+4. `rebootstrapping` must be possible.
+
+### Non-Goals
+1. Revisit how `requestlimit.project.openshift.io/ProjectRequestLimit` works. We want to lift it as is.
+2. Couple a slow moving admission plugin to a fast moving `kube-apiserver`.
+
+### Open Questions
+1. How do we productize the operator? Since this operator is not part of OpenShift core, it can essentially be built and
+ pushed to OperatorHub independently of OpenShift release cycle.
+3. How do we describe version skew limitations to OLM so our operator gets uninstalled *before* an illegal downgrade or
+ upgrade? This is a concrete case of the API we want to use isn't available before 1.16 and after 1.18, the previous API could be gone.
+
+## Proposal
+1. Create a `Validating` admission webhook server that provides `requestlimit.project.openshift.io/ProjectRequestLimit`.
+2. The admission webhook will enable `localhost` serving and `kube-apiserver` will talk to it using a `localhost` URL on the host network.
```

Comment:
have we lost trust in service network and even endpoints so much? @deads2k do we have an endpoint mode for admission webhooks?

Reply:
> have we lost trust in service network and even endpoints so much? @deads2k do we have an endpoint mode for admission webhooks?

For a kube-apiserver admission plugin, it's hard to justify the added cost of the service network.  And for anything controlling pods, we only want to exclude the minimal set.  I think this design is likely to change.

Reply:
> This design has a number of design choices which we would never expect our customers or even other teams to implement. It basically means that we refuse to dogfood what we provide. I am against this in the current form.
> 
> Namely:
> 
> * communication through localhost
> * not using service-ca
> * using indirection through aggregation
> 
> /hold

@sttts we are no longer using localhost serving, so we have dropped localhost access. The webhook will now be fronted by a `Service`. And thus we will use `service-ca` operator to populate certs. 

I think we still want to use aggregation because we will get authentication and authorization for free.

Reply:
yes